Restricted execution of commands to members with `dev` role and changed command/event loading

AidenMontgomery opened the pull request:
**Role Permissions**

It is now possible to specify the roles required to use a command. On execution of a command the executing member's roles are checked against the configured roles list and the command is aborted if they do not meet the requirements.

**Command and Event registration**

I don't really know why, but it felt wrong to be using `fs` to load commands and events when deploying commands and running the bot. They are now loaded with more intention through exported lists of commands and events.

*Commands*
```javascript
export interface ISlashCommandConfig {
  name: string;
  roles: string[];
  commandJSON: () => RESTPostAPIApplicationCommandsJSONBody;
  execute: (interaction: CommandInteraction) => Promise<void>;
}
```

The existing `commands` were moved into a `slashCommands` folder as in the future we may have some more _traditional_ commands, which aren't slash commands.

*Events*
```javascript
export interface IEventConfig<K extends keyof ClientEvents> {
  name: K,
  once: boolean,
  execute: (...args: ClientEvents[K]) => Promise<void>
}
```

The `IEventConfig` interface is keyed to the events that are potentially raised by the Discord Client, which allows the code registering the events to pass the type check for the `execute` function.

crondinini reviewed (approved):
I was also bothered by the use of fs for setting up the commands, so thank you for this PR!

AidenMontgomery committed:
Removed I prefix from interfaces

Removed I prefix from names of interfaces
Removed a loop and made it into a map

Explicity registering the 2 events that we have in the system

AidenMontgomery committed:
Restricted execution of commands to members with `dev` role

Refactored the commands and events registration, so that we no longer rely on loading files from disk and TypeScript accepts all of the typing.

AidenMontgomery committed:
Fixed tests

AidenMontgomery committed:
Devie functionality + Load Commands on start

This commit adds the functionality from the `devie-bot` to the `resource-bot` in preparation to merge the two projects.
This includes adding a new event handler `messageReactionAdd.ts` and associated environment settings.
The new event handler requires access to some `partials` as reactions can be added to old messages which will need to be fetched.
`partials: ['MESSAGE', 'CHANNEL', 'REACTION'],`
These were copied directly from the current `devie-bot` implementation.
It also requires a new *intent* so that it can access _reactions_ added to _guild_messages_.
`Intents.FLAGS.GUILD_MESSAGE_REACTIONS,`

Additionally this commit makes the bot register all commands on startup, so that there is not a separate process for updating the command list.
Loading the commands has been moved into the discord client and will be executed on startup.
The index file was upated to allow the async function to be `await`ed.

AidenMontgomery committed:
Added new env vars to test setup
